bitnami/clickhouse/24/debian-12/rootfs/opt/bitnami/scripts/libclickhouse.sh

```diff
@@ -233,7 +233,7 @@ clickhouse_start_bg() {
     is_clickhouse_running && return
     # This function is meant to be called for internal operations like the init scripts
     local -r cmd=("${CLICKHOUSE_BASE_DIR}/bin/clickhouse-server")
-    local -r args=("--config-file=${CLICKHOUSE_CONF_FILE}" "--pid-file=${CLICKHOUSE_PID_FILE}" "--" "--listen_host=127.0.0.1")
+    local -r args=("--config-file=${CLICKHOUSE_CONF_FILE}" "--pid-file=${CLICKHOUSE_PID_FILE}" "--" "--listen_host=0.0.0.0")
```

Comment:
Thanks for asking, 

As you can see, the initial purpose of running a custom script is to bootstrap ClickHouse configurations before publishing it to other to use (for example my use case requires a lot of Databases and Users to be ready before integrating with other systems). This line of code will never be triggered if user is not meant to use a custom init/start script, so this change means a lots to use cases like mines.

It will not be a good practice to log in to the ClickHouse database and manually create databases/users using command lines because it will take time and is not a very good automated approach (having to manually do such thing in every new deployment seem not right)

My use case also require the ClickHouse setup in a cluster of **3 replicas**, which requires communication between the ClickHouse nodes for setting up the cluster configuration. This is why `listen_host` being set to `127.0.0.1` will fail in setting up the cluster, and the whole deployment will also fail.

The main point here is about the Cluster setup, if the nodes cannot communicate to each other, the setup will fail.

One more thing to notice: When running **/script/setup.sh** on every server start, it has already used the **--listen_host=0.0.0.0** option in [exec /opt/bitnami/scripts/clickhouse/entrypoint.sh /opt/bitnami/scripts/clickhouse/run.sh -- --listen_host=0.0.0.0](https://github.com/bitnami/charts/blob/main/bitnami/clickhouse/templates/scripts-configmap.yaml), so I don't think this small change will break anything.

Thanks & Regards,

Reply:
Thank you @Trungtin1011. Did you explore this other setting? [interserver_listen_host](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#interserver-listen-host). Could it be what we are looking for?

Reply:
Using [interserver_listen_host](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#interserver-listen-host) could be the answer. 

However the document also says that **The default value equals to listen_host setting**. As I want to minimize changes as much as possible because I don't know whether adding a branch new option to the existing command (in this case: `exec /opt/bitnami/scripts/clickhouse/entrypoint.sh /opt/bitnami/scripts/clickhouse/run.sh -- --listen_host=127.0.0.1 --interserver_listen_host=0.0.0.0`) would break the whole configuration or not.

I see that the time span between init/start script is run and the actual server is run is not too big, so I still want to keep it as simple as possible.

In case you don't want to change the `--listen_host` option, please allow me some times to test with adding the option `--interserver_listen_host=0.0.0.0` to the image.

Reply:
Hi @andresbono,

I've just tested the configuration by specifying `local -r args=("--config-file=${CLICKHOUSE_CONF_FILE}" "--pid-file=${CLICKHOUSE_PID_FILE}" "--" "--listen_host=127.0.0.1" "--interserver_listen_host=0.0.0.0")`. The result is the same.

```bash
clickhouse 01:37:20.14 INFO  ==> ** Starting ClickHouse setup **
clickhouse 01:37:20.17 INFO  ==> Copying mounted configuration from /bitnami/clickhouse/etc
clickhouse 01:37:20.19 INFO  ==> Starting ClickHouse in background
clickhouse 01:37:25.20 INFO  ==> ClickHouse started successfully
clickhouse 01:37:25.21 INFO  ==> Loading user's custom files from /docker-entrypoint-initdb.d
clickhouse 01:37:25.21 WARN  ==> Sourcing /docker-entrypoint-initdb.d/platform_init_script.sh as it is not executable by the current user, any error may cause initialization to fail
Received exception from server (version 24.8.4):
Code: 999. DB::Exception: Received from localhost:9000. Coordination::Exception. Coordination::Exception: All connection tries failed while connecting to ZooKeeper. nodes: 10.10.2.33:2181, 10.10.2.174:2181
Poco::Exception. Code: 1000, e.code() = 111, Connection refused (version 24.8.4.13 (official build)), 10.10.2.33:2181
Poco::Exception. Code: 1000, e.code() = 111, Connection refused (version 24.8.4.13 (official build)), 10.10.2.174:2181
Poco::Exception. Code: 1000, e.code() = 111, Connection refused (version 24.8.4.13 (official build)), 10.10.2.33:2181
Poco::Exception. Code: 1000, e.code() = 111, Connection refused (version 24.8.4.13 (official build)), 10.10.2.174:2181
Poco::Exception. Code: 1000, e.code() = 111, Connection refused (version 24.8.4.13 (official build)), 10.10.2.33:2181
Poco::Exception. Code: 1000, e.code() = 111, Connection refused (version 24.8.4.13 (official build)), 10.10.2.174:2181
. (KEEPER_EXCEPTION)
(query: CREATE DATABASE IF NOT EXISTS test ON CLUSTER default;)
```

I had set a command to printout all the logs when ClickHouse is starting in background for troubleshooting and found bellow errors:
1. <Error> virtual bool DB::DDLWorker::initializeMainThread(): Code: 999. Coordination::Exception: All connection tries failed while connecting to ZooKeeper. nodes: 10.10.2.174:2181, 10.10.2.205:2181, 10.10.1.138:2181
2. Poco::Exception. Code: 1000, e.code() = 111, Connection refused (version 24.8.4.13 (official build))
3. Code: 76. DB::Exception: Cannot lock file /opt/bitnami/clickhouse/tmp/clickhouse.pid. Another server instance in same directory is already running.

Look like the ClickHouse background process has not been terminated before the actual ClickHouse process start.

Also in this log, the Keeper nodes failed to connect through port **2181**, so I also explicitly set [interserver_http_port](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#interserver-http-port) and try with `local -r args=("--config-file=${CLICKHOUSE_CONF_FILE}" "--pid-file=${CLICKHOUSE_PID_FILE}" "--" "--listen_host=127.0.0.1" "--interserver_listen_host=0.0.0.0" "--interserver_http_port=2181")`. This time the ClickHouse cannot even be started:

```bash
clickhouse 01:48:50.44 INFO  ==> ** Starting ClickHouse setup **
clickhouse 01:48:50.47 INFO  ==> Copying mounted configuration from /bitnami/clickhouse/etc
clickhouse 01:48:50.48 INFO  ==> Starting ClickHouse in background
clickhouse 01:49:50.52 ERROR ==> ClickHouse failed to start
```

The same result when I use `--interserver_http_port=9444`.

In conclusion, I still think my original opinition will be OK for this.